File: lib/features/orders/domain/entities/order.dart
import 'package:equatable/equatable.dart';

import '../../../shared/domain/enums.dart';

class OrderItem extends Equatable {
  const OrderItem({
    required this.id,
    required this.orderId,
    required this.nameSnapshot,
    required this.unitPriceSnapshot,
    required this.quantity,
    required this.lineTotal,
    this.menuItemId,
    this.note,
  });

  final String id;
  final String orderId;
  final String? menuItemId;
  final String nameSnapshot;
  final double unitPriceSnapshot;
  final int quantity;
  final double lineTotal;
  final String? note;

  @override
  List<Object?> get props => [
        id,
        orderId,
        menuItemId,
        nameSnapshot,
        unitPriceSnapshot,
        quantity,
        lineTotal,
        note,
      ];
}

class Order extends Equatable {
  const Order({
    required this.id,
    required this.sessionId,
    required this.restaurantId,
    required this.tableId,
    required this.status,
    required this.items,
    required this.subtotal,
    required this.tax,
    required this.serviceCharge,
    required this.grandTotal,
    required this.createdAt,
    required this.updatedAt,
    this.notes,
  });

  final String id;
  final String sessionId;
  final String restaurantId;
  final String tableId;
  final OrderStatus status;
  final List<OrderItem> items;
  final double subtotal;
  final double tax;
  final double serviceCharge;
  final double grandTotal;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  bool get isTerminal =>
      status == OrderStatus.completed || status == OrderStatus.cancelled;

  @override
  List<Object?> get props => [
        id,
        sessionId,
        restaurantId,
        tableId,
        status,
        items,
        subtotal,
        tax,
        serviceCharge,
        grandTotal,
        notes,
        createdAt,
        updatedAt,
      ];
}

File: backend/app/Models/Order.php
<?php

namespace App\Models;

use App\Enums\OrderStatus;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Order extends Model
{
    protected $fillable = [
        'dining_session_id',
        'restaurant_id',
        'table_id',
        'status',
        'subtotal',
        'tax',
        'service_charge',
        'grand_total',
        'notes',
        'idempotency_key',
    ];

    protected function casts(): array
    {
        return [
            'status' => OrderStatus::class,
            'subtotal' => 'decimal:2',
            'tax' => 'decimal:2',
            'service_charge' => 'decimal:2',
            'grand_total' => 'decimal:2',
        ];
    }

    public function diningSession(): BelongsTo
    {
        return $this->belongsTo(DiningSession::class);
    }

    public function restaurant(): BelongsTo
    {
        return $this->belongsTo(Restaurant::class);
    }

    public function table(): BelongsTo
    {
        return $this->belongsTo(RestaurantTable::class, 'table_id');
    }

    public function items(): HasMany
    {
        return $this->hasMany(OrderItem::class);
    }
}
